File: Sources/Core/LayoutNode.m
//
//  LayoutNode.m
//  YogaFlexBoxKit
//
//  Created by Apple on 2019/5/21.
//  Copyright © 2019 zjade. All rights reserved.
//

#import "LayoutNode.h"

static YGConfigRef globalConfig;
@interface AbstractLayoutNode()

@end
@implementation AbstractLayoutNode

+ (void)initialize {
    globalConfig = YGConfigNew();
    YGConfigSetExperimentalFeatureEnabled(globalConfig, YGExperimentalFeatureWebFlexBasis, true);
    YGConfigSetPointScaleFactor(globalConfig, [UIScreen mainScreen].scale);
}
- (instancetype)initWithTarget:(id)target {
    if (self = [super init]) {
        _target = target;
        _node = YGNodeNewWithConfig(globalConfig);
        YGNodeSetContext(_node, (__bridge void *) target);
    }
    return self;
}
- (void)dealloc {
    YGNodeFree(self.node);
}
#pragma mark -
- (BOOL)isDirty {
    return YGNodeIsDirty(self.node);
}

- (void)markDirty {
    if (self.isDirty) {
        return;
    }
    YGNodeMarkDirty(self.node);
}
#pragma mark -
- (NSUInteger)numberOfChildren {
    return YGNodeGetChildCount(self.node);
}
-(void)removeAllChildren {
    if (self.node == nil) {
        return;
    }
    YGNodeRemoveAllChildren(self.node);
}
-(void)removeChild:(AbstractLayoutNode*)childNode {
    if (self.node == nil || childNode == nil) {
        return;
    }
    YGNodeRemoveChild(self.node, childNode.node);
}
-(void)insertChild:(AbstractLayoutNode*)childNode index:(uint32_t)index {
    YGNodeInsertChild(self.node, childNode.node, index);
}
-(YGNodeRef __nullable)getChild:(uint32_t)index {
    return YGNodeGetChild(self.node, index);
}

#pragma mark -
-(void)setMeasureFunc:(YGMeasureFunc)func {
    YGNodeSetMeasureFunc(self.node, func);
}
@end

File: Sources/LayoutNode/LayoutNode.swift
//
//  LayoutNode.swift
//  YogaFlexBoxKit
//
//  Created by Apple on 2019/5/22.
//  Copyright © 2019 zjade. All rights reserved.
//

import Foundation

public class LayoutNode: AbstractLayoutNode {
    @available(*, unavailable, message: "使用layoutable")
    public override var target: AnyObject? {
        return super.target
    }
    var layoutable: Layoutable! {
        return super.target as? Layoutable
    }
    @available(*, unavailable, message: "内部使用")
    public override var node: YGNodeRef {
        return super.node
    }
    var yogaNode: YGNodeRef {
        return super.node
    }
    public var isIncludedInLayout: Bool = true
    
    open var isLeaf: Bool {
        return layoutable?.isLeaf ?? true
    }
    
    public override func markDirty() {
        guard isLeaf else { return }
        guard self.isDirty == false else { return }
        /**
         不知道为什么判断有measure函数的时候，再设置一次measure函数。
         通过代码看，应该是measure函数有2种：一种有上下文 一种没有上下文。
         yoga只公开了没有上下文的measure函数，这里判断当之前设置过没有上下文的measure函数，就可以继续设置
         有上下文的度量函数不知道哪里用到了，有可能是保留参数
         */
        if YGNodeHasMeasureFunc(yogaNode) {
            self.setMeasureFunc(measureView)
        }
        super.markDirty()
    }
}
